### DISCUSSION TITLE
DiagnosticsProvider

### DISCUSSION THREAD

**Discussion by ejfasting** [eclipse-langium/langium#1494] [Q&A] [answered]
Hello, 

I have created a language with langium, and am using it from another LSP server. 
It works great to forward requests to langium for completionItems:
`const langiumDocument = shared.workspace.LangiumDocumentFactory.fromTextDocument(textDocument); \n
const params = { textDocument: document, position: position }; \n
const result = await langium.lsp.CompletionProvider?.getCompletion(langiumDocument, params, token);`

But there seems to be no "DiagnosticsProvider", and I was wondering how Langium does this behind the scenes?
After tracing the communication between client and server it looks like the client notifies langium with a 'textDocument/didChange'-event, and then langium does 'magic' and forces a 'textDocument/publishDiagnostics' ? 

So, to summarise, is there a way for langium to give me the Diagnostics, so that i can forward/provide it myself through my own LSP server?

//Eivind

**Comment by msujew** [comment]
Hey @ejfasting,

All instances of `LangiumDocuments` have their respective diagnostics attached to them (in the `diagnostics` field). However, for those to actually be filled, you need to go through the document lifecycle by calling `DocumentBuilder#build` on the document.

Note that Langium uses a few performance optimization internally so that documents don't need to be recomputed on every invocation of an LSP service when running inside of the "normal" language server use case. The `DocumentBuilder` together with the `LangiumDocuments` service (in tandem with a few other services) will cache data on `LangiumDocuments` as to minimize the computations necessary. See also [this documentation page](https://langium.org/docs/reference/document-lifecycle/).

**Reply by ejfasting** [reply, marked as answer]
Thank you for the feedback @msujew !

I actually dug into your sourcecode to get a little more insight, and ended up with using the DocumentValidator.validateDocument(langiumDocument), which gives me the Diagnostics[] (as you also say). 

I'll look at the documentation to fine-tune it.

Thanks again!

//Eivind
